specs/consensus.md

```diff
@@ -595,12 +596,47 @@ At the beginning of the block, rewards are distributed to the block proposer.
 Apply the following to the state:
 
 ```py
+proposer = state.activeValidatorSet[block.header.proposerAddress]
+
 rewardFactor = (TARGET_ANNUAL_ISSUANCE * BLOCK_TIME) / (SECONDS_PER_YEAR * sqrt(GENESIS_COIN_COUNT))
-state.activeValidatorSet[block.header.proposerAddress].pendingRewards += rewardFactor * sqrt(state.activeValidatorSet.votingPower)
+blockReward = rewardFactor * sqrt(state.activeValidatorSet.activeVotingPower)
+state.activeValidatorSet.proposerBlockReward = blockReward
+
+state.activeValidatorSet[block.header.proposerAddress] = proposer
 ```
 
 #### End Block
 
+```py
+account = state.accounts[block.header.proposerAddress]
+
+if account.status == AccountStatus.ValidatorUnbonding
+      account.status == AccountStatus.ValidatorUnbonded
+    proposer = state.inactiveValidatorSet[block.header.proposerAddress]
+else if account.status == AccountStatus.ValidatorBonded
+    proposer = state.activeValidatorSet[block.header.proposerAddress]
+
+blockReward = state.activeValidatorSet.proposerBlockReward
+commissionReward = proposer.commission * blockReward
+proposer.commissionRewards += commissionReward
+proposer.pendingRewards += blockReward - commissionReward
+
+# Even though the voting power hasn't changed yet, we consider this a period change.
+# Note: this isn't perfect because the block reward is shared with delegations
+#  _through_ this block rather than up to the beginning of this block.
+proposer.latestEntry += proposer.pendingRewards // proposer.votingPower
+proposer.pendingRewards = 0
+
+proposer.votingPower += blockReward
+state.activeValidatorSet.activeVotingPower += blockReward
+
+if account.status == AccountStatus.ValidatorUnbonding
+      account.status == AccountStatus.ValidatorUnbonded
+    state.inactiveValidatorSet[block.header.proposerAddress] = proposer
+else if account.status == AccountStatus.ValidatorBonded
+    state.activeValidatorSet[block.header.proposerAddress] = proposer
+```
+
```

Comment:
Is there an equivalent of allocating rewards according to voting power?
e.g. this: https://github.com/cosmos/cosmos-sdk/blob/e2f510afcc3478edab4a7d8d786dccbec333dddd/x/distribution/keeper/allocation.go#L80-L95

Reply:
No, both the block subsidy and fees are given entirely to the proposer. Over the long run this averages out to the same thing as if you spread it out each block so I don't think it makes much difference.

Reply:
That makes sense intuitively and  sounds reasonable but might require further analysis (?). For a first implementation, we will use the distribution module's logic and later change that logic accordingly.